WeakAuras/Transmission.lua

```diff
@@ -730,283 +953,390 @@ local function scamCheck(codes, data)
-    local displayName = regionData and regionData.displayName or regionType or "";
 
-    local tooltip = {
-      {2, data.id, "          ", 0.5333, 0, 1},
-      {2, displayName, "          ", 1, 0.82, 0},
-      {1, " ", 1, 1, 1}
-    };
+function WeakAuras.diff(ours, theirs)
+  -- generates a diff which WeakAuras.Update can use
+  if not ours or not theirs then return end
+  return recurseDiff(ours, theirs)
+end
 
-    local codes = {};
-    local highestVersion = data.internalVersion or 1;
-    scamCheck(codes, data);
-    if(children) then
-      for index, childData in pairs(children) do
-        tinsert(tooltip, {2, " ", childData.id, 1, 1, 1, 1, 1, 1});
-        scamCheck(codes, childData);
-        local childVersion = childData.internalVersion or 1;
-        highestVersion = max(highestVersion, childVersion);
-      end
-      if(#tooltip > 3) then
-        tooltip[4][2] = L["Children:"];
-      else
-        tinsert(tooltip, {1, L["No Children:"], 1, 1, 1});
+function WeakAuras.MatchData(data, children)
+  -- attempts to match the pending import with existing auras
+  -- upon finding a match, will generate a diff as well
+  local oldData, uids = WeakAuras.GetData(data.id), {}
+  if oldData and data.uid ~= nil and oldData.uid ~= data.uid then 
+    oldData = nil 
+  end
+  local match = oldData ~= nil
+  if data.uid then uids[data.uid] = 0 end
+  local oldChildren
+  if children then
+    oldChildren = {}
+    for i, child in ipairs(children) do
+      if child.uid then uids[child.uid] = i end
+      local oldChild = WeakAuras.GetData(child.id)
+      if oldchild and (not child.uid or child.uid == oldChild.uid) then
+        oldChildren[i] = oldChild
       end
-    elseif(data.controlledChildren) then
-      for index, childId in pairs(data.controlledChildren) do
-        tinsert(tooltip, {2, " ", childId, 1, 1, 1, 1, 1, 1});
+    end
+  end
+  for id, existingData in pairs(WeakAurasSaved.displays) do
+    local uindex = existingData.uid and uids[existingData.uid]
+    if uindex == 0 then
+      match = true
+      oldData = existingData
+    elseif uindex then
+      match = true
+      oldChildren[uindex] = existingData
+    end
+  end
+  if match then
+    local info = {
+      unmodified = 0,
+      modified = 0,
+      added = 0,
+      tabsChanged = {}
+    }
+    local diff = WeakAuras.diff(oldData, data)
+    local modified = diff ~= nil
+    local tabsChanged = 0
+    if modified then
+      info.modified = 1
+      for key in pairs(diff) do
+        local button = keyToButton[key]
+        if button and not info.tabsChanged[button] then
+          info.tabsChanged[button] = true
+          tabsChanged = tabsChanged + 1
+        end
+        if tabsChanged >= 7 then break end
       end
-      if(#tooltip > 3) then
-        tooltip[4][2] = L["Children:"];
-      else
-        tinsert(tooltip, {1, L["No Children:"], 1, 1, 1});
+    end
+    if oldData then
+      info[0] = {oldData, diff}
```

Comment:
Hmmmm, I prefer info.oldData[0], info.diff[0] = oldData, diff. As I think that leads to more readable code in other places.

Reply:
That’s fine. I’ll rearrange it.

Reply:
That’s fine. I’ll rearrange it.